Enhanced redeem with the signature of the issuer

barvhaim opened the pull request.

alexandrosfilios commented on Makefile:
```diff
@@ -148,4 +148,8 @@ tokengen:
 txgen:
-	@go install github.com/IBM/idemix/tools/idemixgen
\ No newline at end of file
+	@go install github.com/IBM/idemix/tools/idemixgen
```
I think you will need a specific version here. the problem is that the dependency sometimes is updated because of other dependencies during a go mod. then you have to reset it back to that specific version needed.

barvhaim replied:
This line stayed as-is and out of the scope of this PR

alexandrosfilios commented on integration/ports.go:
```diff
@@ -60,8 +60,8 @@ var (
 		WebSocketNoReplication,
-		LibP2PNoReplication,
-		WebSocketWithReplication,
+		//LibP2PNoReplication,
```
not to forget

aaadir replied:
done

barvhaim commented:
@aaadir I implemented the fabotken + zkat logic, fixed the failing unit-tests and most of the integration tests.. what is left is integration tests related to "Auditor as an Issuer" cases

adecaro commented on token/core/fabtoken/v1/actions/transfer_test.go:
```diff
@@ -255,6 +255,7 @@ func TestTransferAction_Serialization(t *testing.T) {
 		Metadata: map[string][]byte{
 			"metadata": []byte("{\"foo\":\"bar\"}"),
 		},
+		Issuer: nil,
```
Please, a value different from nil to make sure the serialization works as expected.

aaadir replied:
The Issuer field should be nil for non-Redeem transactions.
For redeem it will be populated to the issuer's-PK inside the action.Serialize() that follows immediately after this line.

adecaro commented on token/core/fabtoken/v1/actions/transfer.go:
```diff
@@ -164,6 +167,11 @@ func (t *TransferAction) GetMetadata() map[string][]byte {
 }
 
+// GetIssuer returns the issuer to sign the transaction
+func (t *TransferAction) GetIssuer() driver.Identity {
```
Please, add a unit-test for this function.

aaadir replied:
Added TestTransferAction_GetIssuer to token/core/fabtoken/v1/actions/transfer_test.go
done.

adecaro commented on token/core/fabtoken/v1/setup/setup_test.go:
```diff
@@ -10,6 +10,7 @@ import (
 	"testing"
 
 	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+
```
remove this line

aaadir replied:
done

adecaro commented on token/core/fabtoken/v1/transfer.go:
```diff
@@ -46,6 +46,7 @@ func NewTransferService(
 // Transfer returns a TransferAction as a function of the passed arguments
 // It also returns the corresponding TransferMetadata
 func (s *TransferService) Transfer(ctx context.Context, _ string, _ driver.OwnerWallet, tokenIDs []*token.ID, Outputs []*token.Token, opts *driver.TransferOptions) (driver.TransferAction, *driver.TransferMetadata, error) {
+	var isRedeem bool
```
please, move this declaration closer to where it is needed.

aaadir replied:
it's indeed needed much later, but it's set in the iteration over the outputs here below. But I moved the var declaration to right before this iteration.
done.

adecaro commented on token/core/fabtoken/v1/validator/validator_transfer.go:
```diff
@@ -30,6 +30,7 @@ func TransferSignatureValidate(ctx *Context) error {
 		return errors.Errorf("invalid number of token inputs, expected at least 1")
 	}
 
+	var isRedeem bool
```
please, move this declaration closer to where it is needed.

aaadir replied:
moved to before the place it is set, while iterating the outputs. 
done.

adecaro commented on token/core/zkatdlog/nogh/v1/transfer/action.go:
```diff
@@ -243,6 +247,11 @@ func (t *Action) GetMetadata() map[string][]byte {
 }
 
+// GetIssuer returns the issuer to sign the transaction
+func (t *Action) GetIssuer() driver.Identity {
```
please, add a unit-tests for this function.

aaadir replied:
done

adecaro commented on token/core/zkatdlog/nogh/v1/transfer/action.go:
```diff
@@ -313,6 +322,13 @@ func (t *Action) Serialize() ([]byte, error) {
 		return nil, errors.Wrap(err, "failed to serialize outputs")
 	}
 
+	var issuer *pp.Identity
```
please, update TestAction_Validate to check also the added code.

aaadir replied:
TestAction_Validate  doesn't really check the updated Serialize code, just runs action.Validate().
But I added an issuer to the action anyway, to see that Validate works.
Serialize is actually tested in another test here (TestSerialization) so I also updated it (by creating an issuer in randomAction).
done.

adecaro commented on token/core/zkatdlog/nogh/v1/validator/validator_transfer.go:
```diff
@@ -56,6 +57,33 @@ func TransferSignatureValidate(ctx *Context) error {
+	}
+
+	if isRedeem {
+		ctx.Logger.Infof("action is a redeem, verify the signature of the issuer")
```
please, use `Debugf` here

adecaro commented on token/driver/request.go:
```diff
@@ -373,6 +373,8 @@ type TransferMetadata struct {
 	// but needs to sign the request
 	ExtraSigners []Identity
+	// Issuer contains the identity of the issuer to sign the transfer action
+	Issuer Identity
```
please, update TestTokenRequestMetadataSerialization, to include also this field

aaadir replied:
done

adecaro commented on token/services/ttx/transfer_opts.go:
```diff
@@ -0,0 +1,44 @@
+	IssuerFSCIdentityKey = "IssuerFSCIdentityKey"
+)
+
+func WithFSCIssuerIdentity(issuerFSCIdentity view.Identity) token.TransferOption {
```
please, add comments to these two functions

aaadir replied:
done

adecaro reviewed (changes requested):
LGTM overall modulo the requested changes. Thanks

adecaro reviewed (approved):
LGTM
